Add resource metric integration test

varunch77 opened the pull request:
# Description of the issue
There is no integration test that verifies that an entity is being sent when we send resource metrics such as cpu and memory to CloudWatch—this is currently a gap in our testing.

# Description of changes
- Create an agent config to send resource metrics to Cloudwatch
- Sign and send a request that calls `ListEntitiesForMetric` to get the entity associated with our metrics
- Validate the returned entity
- Update our integration testing IAM role to include permission for `ListEntitiesForMetric`

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
Integration test run: https://github.com/aws/amazon-cloudwatch-agent/actions/runs/12838570212

varunch77 commented on test/e2e/jmx/jmx_test.go:
unrelated to code, just a linter change

varunch77 commented on test/e2e/jmx/jmx_test.go:
```diff
@@ -25,7 +25,7 @@ import (
 
 var (
-	env       *environment.MetaData
+	env *environment.MetaData
```
unrelated to code, just a linter change

lisguo commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -274,6 +282,103 @@ func TestLogGroupClass(t *testing.T) {
+	common.StopAgent()
+
+	// this section builds, signs, and sends the request
+	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion("us-west-2"))
```
nit: define const for region

varunch77 replied:
Addressed

https://github.com/aws/amazon-cloudwatch-agent-test/blob/1c5bba211f1aedc86e2b3a56134478dd0ad1967d/test/entity/entity_test.go#L26

lisguo commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -274,6 +282,103 @@ func TestLogGroupClass(t *testing.T) {
+	//   }'
+
+	// build the request
+	req, err := http.NewRequest("POST", "https://monitoring.us-west-2.amazonaws.com/", bytes.NewReader(body))
```
Can we extract this out into a helper func? Maybe a `getEntitiesForMetric(body)` function that returns the list of entities for the given metric?

varunch77 replied:
Added `BuildListEntitiesForMetricRequest` in `util/common/metrics.go`

lisguo commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -274,6 +282,103 @@ func TestLogGroupClass(t *testing.T) {
 	}
 }
 
+func TestResourceMetrics(t *testing.T) {
```
The file name is called "publish_logs" but aren't we publishing metrics?

varunch77 replied:
Removed logs

lisguo commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -274,6 +282,103 @@ func TestLogGroupClass(t *testing.T) {
+
+	defer logFile.Close()
+	defer os.Remove(logFilePath)
+	defer awsservice.DeleteLogGroupAndStream(instanceId, instanceId)
```
we shouldn't have to delete any log groups if we are just publishing a resource metric

lisguo commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -274,6 +282,103 @@ func TestLogGroupClass(t *testing.T) {
+	defer resp.Body.Close()
+
+	// parse and verify the response
+	var response struct {
```
Similar to the above, can we have a helper func `validateEntities(expectedEntities, actualEntities)` so that we can re-use these functions?

Even better, we can put these in some util package so that we can use it in multiple tests

varunch77 replied:
Addressed, used existing helper functions in the `entity_test.go` file

lisguo commented on test/cloudwatchlogs/resources/config_log_resource.json:
```diff
@@ -0,0 +1,46 @@
+            "InstanceId": "${aws:InstanceId}"
+        }
+    },
+    "logs": {
```
still confused...I thought this PR was just for metrics so we shouldn't need the logs section?

varunch77 commented:
Changes:
- Removed anything logs-related
- Moved the location of the test into the entity folder
- Created a helper function in the util/common/metrics package to handle building the `ListEntitiesForMetric` call
- Updated the entity validation to use existing helper functions

varunch77 commented on test/cloudwatchlogs/publish_logs_test.go:
```diff
@@ -274,6 +282,103 @@ func TestLogGroupClass(t *testing.T) {
+	// start agent and write metrics and logs
+	common.CopyFile(configPath, configOutputPath)
+	common.StartAgent(configOutputPath, true, false)
+	time.Sleep(2 * time.Minute)
```
Addressed https://github.com/aws/amazon-cloudwatch-agent-test/blob/1c5bba211f1aedc86e2b3a56134478dd0ad1967d/test/entity/entity_test.go#L250

zhihonl commented on test/entity/entity_test.go:
```diff
@@ -237,6 +241,62 @@ func TestPutLogEventEntityEKS(t *testing.T) {
+	common.StopAgent()
+
+	// build ListEntitiesForMetric request:
+	requestBody := []byte(fmt.Sprintf(`{
```
nit: I think would be nice to turn these into test case input, then in the future if someone else to implement a new test, they can just write an additional case instead of copying this entire boilerplate code. For example: https://github.com/aws/amazon-cloudwatch-agent-test/pull/453/files#diff-13510c6aa10605c1f76162fde322c3a4da1cf75441a3a93cb340fbb9d36c480dR186

zhihonl commented:
Can you attach a new integration test run with this change so we can make sure the test still passes?

varunch77 commented:
New integration test run: https://github.com/aws/amazon-cloudwatch-agent/actions/runs/12994059613

Example of `EntityMetrics` test passing: https://github.com/aws/amazon-cloudwatch-agent/actions/runs/12994059613/job/36237893873#step:8:4090

lisguo commented on test/metric_value_benchmark/entity_metrics_test.go:
```diff
@@ -0,0 +1,139 @@
+	}
+
+	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
+		return testResult
```
nit: Each time the test run fails, we should log the reason why so that we can troubleshoot better

varunch77 replied:
Addressed
